### PR TITLE
add scaling threshold to crd

### PR DESCRIPTION
**Pull request by OliverMKing**
# Description

adds scaling threshold to crd to configure how fast hpa scales. The options provided are known good values. We use enums to control the actual values which we scale test internally. We control cpu requests so that is another reason why these are enums.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

locally, e2e, unit

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by sabbour — DISMISSED** [review]
Approved from naming/description perspective

File: api/v1alpha1/nginxingresscontroller_types.go [hunk 1]
```diff
@@ -91,8 +91,24 @@ type Scaling struct {
 	// +kubebuilder:validation:Minimum=1
 	// +optional
 	MaxReplicas *int32 `json:"maxReplicas,omitempty"`
+
+	// Threshold defines how quickly the Ingress Controller pods should scale based on workload. Rapid means the Ingress Controller
+	// will scale quickly and aggressively, which is the best choice for handling sudden and significant traffic spikes. Steady
+	// is the opposite, prioritizing cost-effectiveness. Steady is the best choice when fewer replicas handling more work is desired or when
+	// traffic isn't expected to fluctuate. Balanced is a good mix between the two that works for most use-cases. It defaults to balanced.
```
Comment (jaiveerk) [inline api/v1alpha1/nginxingresscontroller_types.go:98]: nit: "It defaults to balanced" is a bit unclear - can we make this Threshold defaults to balanced or "... that forks for most use-cases and is the default value"

File: pkg/controller/nginxingress/nginx_ingress_controller.go [hunk 2]
```diff
@@ -560,3 +567,31 @@ func ToNginxIngressConfig(nic *approutingv1alpha1.NginxIngressController, defaul
 
 	return nginxIng
 }
+
+func getTargetCPUUtilizationPercentage(nic *approutingv1alpha1.NginxIngressController) int32 {
+	if nic == nil {
+		return defaultTargetCPUUtilization
+	}
+
+	scaling := nic.Spec.Scaling
```
Comment (jaiveerk) [inline pkg/controller/nginxingress/nginx_ingress_controller.go:576]: is it possible for nic.Spec to be nil/do we need a nil check for that as well

File: pkg/controller/nginxingress/nginx_ingress_controller.go [hunk 3]
```diff
@@ -35,6 +35,12 @@ const (
 	defaultMaxReplicas = 100
 )
 
+const (
+	rapidTargetCpuUtilization   = 55
+	defaultTargetCPUUtilization = 70
```
Comment (jaiveerk) [inline pkg/controller/nginxingress/nginx_ingress_controller.go:40]: nit: can we rename to balanced? I don’t want to couple the fact that we have balanced as an option with the fact that we also have a default. Instead, we should have logic that, in the default case, returns balanced (or whatever we want it to) - not define balanced as the default to begin with

File: pkg/controller/nginxingress/nginx_ingress_controller_test.go [hunk 4]
```diff
@@ -1099,12 +1111,61 @@ func TestToNginxIngressConfig(t *testing.T) {
...
+			},
+			want: manifests.NginxIngressConfig{
+				ControllerClass:                "approuting.kubernetes.azure.com/nicName",
+				ResourceName:                   "controllerNamePrefix-0",
+				ServiceConfig:                  &manifests.ServiceConfig{},
+				IcName:                         "ingressClassName",
+				MaxReplicas:                    defaultMaxReplicas,
+				MinReplicas:                    defaultMinReplicas,
+				TargetCPUUtilizationPercentage: steadyTargetCPUUtilization,
 			},
```
Comment (jaiveerk) [inline pkg/controller/nginxingress/nginx_ingress_controller_test.go:1169]: Can you add a custom fields with balanced threshold case as well

**Review by jaiveerk — CHANGES_REQUESTED** [review]
few changes requested. I know we’ve got fixture tests so our bases should be covered but is there any feasible and useful way we can test for this in e2e to ensure our output values are correct all the way through to hpa deployment?

File: pkg/controller/nginxingress/nginx_ingress_controller.go (hunk 2 above)
Comment (OliverMKing) [reply pkg/controller/nginxingress/nginx_ingress_controller.go:576]: no

File: api/v1alpha1/nginxingresscontroller_types.go (hunk 1 above)
Comment (OliverMKing) [reply api/v1alpha1/nginxingresscontroller_types.go]: done

File: pkg/controller/nginxingress/nginx_ingress_controller_test.go (hunk 4 above)
Comment (OliverMKing) [reply pkg/controller/nginxingress/nginx_ingress_controller_test.go:1193]: done

**Review by jaiveerk — APPROVED** [review]

**Review by bfoley13 — APPROVED** [review]
Changes look good

**Pull request merged**